pkg/autodiscovery/listeners/kube_services.go

```diff
@@ -225,14 +265,232 @@ func (l *KubeServiceListener) removeService(ksvc *v1.Service) {
+		l.createEndpoint(castedObj, false)
+		return
+	}
+	if endpointsDiffer(castedObj, castedOld) {
+		l.removeEndpoint(castedObj)
+		l.createEndpoint(castedObj, false)
+	}
+}
+
+// endpointsDiffer compares two endpoints to only go forward
+// when relevant fields are changed. This logic must be
+// updated if more fields are used.
+func endpointsDiffer(first, second *v1.Endpoints) bool {
+	// Quick exit if resversion did not change
+	if first.ResourceVersion == second.ResourceVersion {
+		return false
+	}
+
+	// Subsets
+	if len(first.Subsets) != len(second.Subsets) {
+		return true
+	}
+
+	// Addresses and Ports
+	for i := range first.Subsets {
+		if len(first.Subsets[i].Addresses) != len(second.Subsets[i].Addresses) {
+			return true
+		}
+		if len(first.Subsets[i].Ports) != len(second.Subsets[i].Ports) {
+			return true
+		}
+		// Addresses
+		for j := range first.Subsets[i].Addresses {
+			if first.Subsets[i].Addresses[j].IP != second.Subsets[i].Addresses[j].IP {
+				return true
+			}
+			if first.Subsets[i].Addresses[j].Hostname != second.Subsets[i].Addresses[j].Hostname {
+				return true
+			}
+		}
+		// Ports
+		for j := range first.Subsets[i].Ports {
+			if first.Subsets[i].Ports[j].Port != second.Subsets[i].Ports[j].Port {
+				return true
+			}
+			if first.Subsets[i].Ports[j].Name != second.Subsets[i].Ports[j].Name {
+				return true
+			}
+		}
+	}
+	// No relevant change
+	return false
+}
+
+func (l *KubeServiceListener) createEndpoint(kendpt *v1.Endpoints, firstRun bool) {
+	if kendpt == nil {
+		return
+	}
+	endptId := fmt.Sprintf("%s/%s", kendpt.Namespace, kendpt.Name)
+
+	l.m.Lock()
+	for id, isAnnotated := range l.endpointsAnnotated {
+		if id == endptId && isAnnotated {
+			endpts := processEndpoint(kendpt, firstRun)
+			newEndpts, removedEndpts := diffEndpoints(endpts, l.endpoints[endptId])
+			l.updateEndpoints(newEndpts, removedEndpts, endptId)
+			for _, endpt := range newEndpts {
+				log.Debugf("sending new endpoint: %v", endpt)
+				l.newService <- endpt
+			}
+			for _, endpt := range removedEndpts {
+				log.Debugf("removing endpoint: %v", endpt)
+				l.delService <- endpt
+			}
+		}
+	}
+	l.m.Unlock()
+}
+
+func processEndpoint(kendpt *v1.Endpoints, firstRun bool) []*KubeEndpointService {
+	var newEndpointServices []*KubeEndpointService
+
+	// Hosts and Ports
+	var hosts = make(map[string]string)
+	var ports []ContainerPort
+	for i := range kendpt.Subsets {
+		// Hosts
+		for _, host := range kendpt.Subsets[i].Addresses {
+			hosts[host.IP] = host.Hostname
+		}
+		// Ports
+		for _, port := range kendpt.Subsets[i].Ports {
+			ports = append(ports, ContainerPort{int(port.Port), port.Name})
+		}
+	}
+
+	for ip, host := range hosts {
+		// create a separate AD service per host
+		endpt := &KubeEndpointService{
+			entity:       apiserver.EntityForEndpoints(kendpt.Namespace, kendpt.Name, ip),
+			creationTime: integration.After,
+			hosts:        map[string]string{host: ip},
```

Comment:
The map key here should be predictable, as it's intended to be used as token in the extended `%%host_KEY%%` variable. In the docker listener, the map key is the docker network name, so we can do `%%host_overlay%%` for the IP on the overlay network.

Here, as there is only one IP per address, you can use "service" as a placeholder. But indeed, the `Hostname` field (if not empty) could be used as a `%%hostname%%` variable value. I missed it as it's optional and not filled for pod-to-service matchings


We can also merge the two loops:
```
for each subset:
    new port map
    for each port:
         fill port map
    for each address:
         generate a service (using subset-local port map)
```